### PR TITLE
fix: Correct Tax Breakup for different tax rates for same hsn code

### PR DESCRIPTION
**Pull request by DaizyModi**
**Issue**
In case of different tax rates for same HSN code, tax rate and taxable amount was getting overriden.

**Before**
image

**After**

image

### CODE REVIEW

File: erpnext/controllers/taxes_and_totals.py [hunk 1]
```diff
@@ -1001,11 +1000,23 @@ def get_itemised_tax_breakup_header(item_doctype, tax_accounts):
...
+
+def _get_itemised_tax_breakup_data(doc, with_tax_account=False):
+	itemised_tax = get_itemised_tax(doc.taxes, with_tax_account=with_tax_account)
 
 	itemised_taxable_amount = get_itemised_taxable_amount(doc.items)
 
-	return itemised_tax, itemised_taxable_amount
+	itemised_tax_data = []
+	for item_code, taxes in itemised_tax.items():
+		for _item_code, taxable_amount in itemised_taxable_amount.items():
```
Comment (vorasmit) [inline erpnext/controllers/taxes_and_totals.py]: Is this loop needed? how about taxable_amount.get(item_code)

File: erpnext/controllers/taxes_and_totals.py [hunk 2]
```diff
@@ -1001,11 +1000,23 @@ def get_itemised_tax_breakup_header(item_doctype, tax_accounts):
 
 @erpnext.allow_regional
 def get_itemised_tax_breakup_data(doc):
-	itemised_tax = get_itemised_tax(doc.taxes)
+	return _get_itemised_tax_breakup_data(doc)
+
+
+def _get_itemised_tax_breakup_data(doc, with_tax_account=False):
```
Comment (vorasmit) [inline erpnext/controllers/taxes_and_totals.py]: No need for a separate function here.

File: erpnext/controllers/taxes_and_totals.py [hunk 3]
```diff
@@ -1048,11 +1059,12 @@ def get_itemised_taxable_amount(items):
...
 
-def get_rounded_tax_amount(itemised_tax, precision):
+def get_rounded_tax_amount(itemised_tax, precision, tax_accounts):
 	# Rounding based on tax_amount precision
-	for taxes in itemised_tax.values():
-		for tax_account in taxes:
-			taxes[tax_account]["tax_amount"] = flt(taxes[tax_account]["tax_amount"], precision)
+	for _itemised_tax in itemised_tax:
+		for key, value in _itemised_tax.items():
+			if key in tax_accounts:
```
Comment (vorasmit) [inline erpnext/controllers/taxes_and_totals.py]: Instead how about checking type of value? Is tax_accounts really needed?

File: erpnext/controllers/taxes_and_totals.py [hunk 4]
```diff
@@ -1048,11 +1059,12 @@ def get_itemised_taxable_amount(items):
...
 
 
-def get_rounded_tax_amount(itemised_tax, precision):
+def get_rounded_tax_amount(itemised_tax, precision, tax_accounts):
 	# Rounding based on tax_amount precision
-	for taxes in itemised_tax.values():
-		for tax_account in taxes:
-			taxes[tax_account]["tax_amount"] = flt(taxes[tax_account]["tax_amount"], precision)
+	for _itemised_tax in itemised_tax:
+		for key, value in _itemised_tax.items():
```
Comment (vorasmit) [inline erpnext/controllers/taxes_and_totals.py]:
```suggestion
	for taxes in itemised_tax:
		for row in taxes.values():
```

**Review by deepeshgarg007 — APPROVED** [review]

**Comment by precihole** [comment]
Hello,

We are getting issue because of this PR in Purchase Order when we do **Get Items From Supplier Quotation**.
Kindly resolve this on urgent basis.
@deepeshgarg007 @DaizyModi @vorasmit @sagarvora 


image

### App Versions
```
{
	"erpnext": "14.32.1",
	"erpnext_telegram_integration": "1.2.0",
	"frappe": "14.42.0",
	"hrms": "14.8.0",
	"india_compliance": "14.13.0",
	"insights": "0.10.0",
	"pl_accounts": "0.0.1",
	"preciholesports": "0.0.1"
}
```
### Route
```
Form/Purchase Order/new-purchase-order-1
```
### Traceback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 94, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 47, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 85, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1619, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/model/mapper.py", line 53, in map_docs
    target_doc = method(*_args)
  File "apps/erpnext/erpnext/buying/doctype/supplier_quotation/supplier_quotation.py", line 142, in make_purchase_order
    doclist = get_mapped_doc(
  File "apps/frappe/frappe/model/mapper.py", line 148, in get_mapped_doc
    postprocess(source_doc, target_doc)
  File "apps/erpnext/erpnext/buying/doctype/supplier_quotation/supplier_quotation.py", line 137, in set_missing_values
    target.run_method("calculate_taxes_and_totals")
  File "apps/frappe/frappe/model/document.py", line 914, in run_method
    out = Document.hook(fn)(self, *args, **kwargs)
  File "apps/frappe/frappe/model/document.py", line 1267, in composer
    return composed(self, method, *args, **kwargs)
  File "apps/frappe/frappe/model/document.py", line 1249, in runner
    add_to_return_value(self, fn(self, *args, **kwargs))
  File "apps/frappe/frappe/model/document.py", line 911, in fn
    return method_object(*args, **kwargs)
  File "apps/erpnext/erpnext/controllers/accounts_controller.py", line 348, in calculate_taxes_and_totals
    calculate_taxes_and_totals(self)
  File "apps/erpnext/erpnext/controllers/taxes_and_totals.py", line 32, in __init__
    self.calculate()
  File "apps/erpnext/erpnext/controllers/taxes_and_totals.py", line 62, in calculate
    self.set_item_wise_tax_breakup()
  File "apps/erpnext/erpnext/controllers/taxes_and_totals.py", line 912, in set_item_wise_tax_breakup
    self.doc.other_charges_calculation = get_itemised_tax_breakup_html(self.doc)
  File "apps/erpnext/erpnext/controllers/taxes_and_totals.py", line 958, in get_itemised_tax_breakup_html
    get_rounded_tax_amount(itemised_tax_data, doc.precision("tax_amount", "taxes"))
  File "apps/erpnext/erpnext/controllers/taxes_and_totals.py", line 1056, in get_rounded_tax_amount
    if isinstance(row, dict) and isinstance(row["tax_amount"], float):
KeyError: 'tax_amount'

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"method": "erpnext.buying.doctype.supplier_quotation.supplier_quotation.make_purchase_order",
		"source_names": "[\"PUR-GR2324-0214\"]",
		"target_doc": "{\"docstatus\":0,\"doctype\":\"Purchase Order\",\"name\":\"new-purchase-order-1\",\"__islocal\":1,\"__unsaved\":1,\"owner\":\"rehan@preciholesports.com\",\"title\":\"{supplier_name}\",\"naming_series\":\"PO-JW.FY.-.####\",\"material_request_type\":\"\",\"transaction_date\":\"2023-07-26\",\"company\":\"Precihole Sports Pvt. Ltd.\",\"apply_tds\":0,\"is_reverse_charge\":0,\"purpose\":\"\",\"is_subcontracted\":0,\"currency\":\"INR\",\"buying_price_list\":\"Supplier Rates\",\"price_list_currency\":\"INR\",\"ignore_pricing_rule\":0,\"set_warehouse\":\"Stores A197 (Material In) - PSPL\",\"tax_withholding_net_total\":0,\"set_reserve_warehouse\":\"Reserve - PSPL\",\"disable_rounded_total\":0,\"apply_discount_on\":\"Grand Total\",\"status\":\"Draft\",\"letter_head\":\"No Letterhead\",\"group_same_items\":0,\"is_internal_supplier\":0,\"party_account_currency\":\"INR\",\"is_old_subcontracting_flow\":0,\"items\":[],\"workflow_state\":\"Draft\",\"advance_paid\":0,\"conversion_rate\":1,\"plc_conversion_rate\":1,\"shipping_address\":\"PSPL Warehouse - 197 - Shipping\",\"shipping_address_display\":\"Plot No. A-197, Road No 16A, Wagle Estate, <br>Thane (W)  Maharashtra 400604<br>Thane,PIN: 400604,<br>Maharashtra, State Code: 27,India<br>\\n<b>Phone:</b> 8879320376 / 77&ensp;<br><b>Email:</b> info@preciholesports.com<br><b>GSTIN:</b> 27AAGCP5774K1ZT\",\"destination\":\"Thane\",\"base_net_total\":0,\"net_total\":0,\"base_total\":0,\"total\":0,\"total_qty\":0,\"rounding_adjustment\":0,\"grand_total\":0,\"taxes_and_charges_deducted\":0,\"taxes_and_charges_added\":0,\"base_grand_total\":0,\"base_taxes_and_charges_added\":0,\"base_taxes_and_charges_deducted\":0,\"total_taxes_and_charges\":0,\"base_total_taxes_and_charges\":0,\"base_rounding_adjustment\":0,\"rounded_total\":0,\"base_rounded_total\":0,\"in_words\":\"\",\"base_in_words\":\"\",\"base_discount_amount\":0}"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.model.mapper.map_docs"
}
```
### Response Data
```
{
	"exception": "KeyError: 'tax_amount'"
}
```

**Comment by precihole** [comment]
Hello,

Our ERP Site is stuck becuase of this what to do for now? @vorasmit

**Comment by vorasmit** [comment]
Hello,

I request you to upgrade the version of India Compliance to v.14.14.0 and the issue will be resolved for you.

**Pull request merged**